onmetal-controller-manager/controllers/networking/natgateway_controller.go

```diff
@@ -0,0 +1,318 @@
+		return r.delete(ctx, log, natGateway)
+	}
+	return r.reconcile(ctx, log, natGateway)
+}
+
+func (r *NatGatewayReconciler) delete(ctx context.Context, log logr.Logger, natGateway *networkingv1alpha1.NATGateway) (ctrl.Result, error) {
+	return ctrl.Result{}, nil
+}
+
+func hash(ip *networkingv1alpha1.NATGatewayDestinationIP) string {
+	return fmt.Sprintf("%s/%d/%d", ip.IP, ip.Port, ip.EndPort)
+}
+
+func (r *NatGatewayReconciler) assignPorts(ctx context.Context, log logr.Logger, natGateway *networkingv1alpha1.NATGateway, destinations map[types.UID]*networkingv1alpha1.NATGatewayDestination, slotsPerIP int, portsPerNetworkInterface int32) ([]networkingv1alpha1.NATGatewayDestination, int32, error) {
+	natGatewayRouting := &networkingv1alpha1.NATGatewayRouting{}
+	if err := r.Get(ctx, types.NamespacedName{Namespace: natGateway.Namespace, Name: natGateway.Name}, natGatewayRouting); client.IgnoreNotFound(err) != nil {
+		return nil, 0, fmt.Errorf("unable to get natgateway routing %s: %w", client.ObjectKeyFromObject(natGatewayRouting), err)
+	}
+
+	used := make(map[string]bool)
+	for _, dest := range natGatewayRouting.Destinations {
+		if v, found := destinations[dest.UID]; found {
+			v.IPs = dest.IPs
+			for _, ip := range v.IPs {
+				used[hash(&ip)] = true
+			}
+		}
+	}
+
+	var (
+		currentIp, currentPort int32 = 0, 0
+		newDestinations        []networkingv1alpha1.NATGatewayDestination
+	)
+	for _, v := range destinations {
+		log.V(2).Info("Start processing destination", "destination", v.Name)
+		if len(v.IPs) != 0 {
+			log.V(2).Info("keep previous port assignment of destination", "destination", v.Name)
+			newDestinations = append(newDestinations, *v)
+			continue
+		}
+
+		log.V(2).Info("find free port for destination", "destination", v.Name)
+		var candidate *networkingv1alpha1.NATGatewayDestinationIP
+		for {
+			if int(currentIp) >= len(natGateway.Status.IPs) {
+				log.V(1).Info("No ports left: Skipping destination", "name", v.Name, "UID", v.UID)
+				r.Eventf(natGateway, corev1.EventTypeWarning, events.ErrorNoPortsLeft, "no ports left: skipping destination %s %s", v.Name, v.UID)
+				candidate = nil
+				break
+			}
+
+			port := currentPort*portsPerNetworkInterface + MIN_PORT
+			candidate = &networkingv1alpha1.NATGatewayDestinationIP{
+				IP:      natGateway.Status.IPs[currentIp].IP,
+				Port:    port,
+				EndPort: port + portsPerNetworkInterface - 1,
+			}
+
+			if _, allocated := used[hash(candidate)]; !allocated {
+				used[hash(candidate)] = true
+				log.V(2).Info("found port range for destination", "destination", v.Name, "start", candidate.Port, "end", candidate.EndPort, "ip", candidate.IP.String())
+				break
+			}
+
+			currentPort++
+			if int(currentPort) > slotsPerIP {
+				currentPort = 0
+				currentIp++
+			}
+		}
+
+		if candidate != nil {
+			v.IPs = []networkingv1alpha1.NATGatewayDestinationIP{*candidate}
+		}
+		newDestinations = append(newDestinations, *v)
+	}
+
+	return newDestinations, int32(len(used)), nil
```

Comment:
We should not report any destination where `IPs` are empty - this would just be unnecessary noise.